### PR TITLE
A metrics adapter for Alibaba Cloud Monitor

### PR DESCRIPTION
**Pull request by osswangxining**
Currently our potentail Istio users are finding how to use existing Aibaba Cloud Monitor instance together within Istio, just likw the AWS Cloud Watch adapter within Istio.

### CODE REVIEW

**Comment by geeknoid** [comment]
Hi there,

Thanks for the new adapter. Two quick comments before a full review:

- Could you redo your PR please? It currently claims there are >5000 changes files, so clearly something is wrong.

- Please run bin/fmt.sh which will reformat your code to comply with our conventions.

Thanks.

**Comment by douglas-reid** [comment]
Can we add more to the PR description around the vendor changes in this PR (which repo and release this includes)?  That would help speed things up (and clarify why there are >5k files changed).

**Comment by osswangxining** [comment]
@douglas-reid  This PR uses the Alibaba Cloud SDK go version as vendor dependencies. You can find from the repo: https://github.com/aliyun/alibaba-cloud-sdk-go

File: mixer/adapter/cloudmonitor/config/config.proto [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY Type, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+// $title: CloudMonitor
+// $description: Adapter for cloudmonitor metrics.
+// $location: https://istio.io/docs/reference/config/policy-and-telemetry/adapters/cloudmonitot.html
```
Comment (venilnoronha) [inline mixer/adapter/cloudmonitor/config/config.proto]: `cloudmonitot.html` -> `cloudmonitor.html`?

**Comment by geeknoid** [comment]
To fix the lint errors, I think you just need to rebase and push again. The errors were elsewhere and have been fixed.

Before embarking on the code review, I need to point out that we're in the process of transitioning over to our [out-of-process adapter model](https://github.com/istio/istio/wiki/Mixer-Out-Of-Process-Adapter-Dev-Guide). Since this is new code, do you think it would be possible to recast this adapter as an out-of-process adapter? You should be able to turn this adapter into an out-of-proc adapter with a relatively simple gRPC shim to sit in front of the code?

Making it an out-of-proc adapter would make it so you could host this code in your own repo, as the adapter would not need to be built into the Mixer binary.

Over the coming months, we will convert most existing adapters into out-of-proc adapters. The only ones left in Mixer will be the trivial ones, like the list checker and the denials adapter. 

CC @mandarjog whose working on out-of-proc adapters.

**Comment by mandarjog** [comment]
We are requesting all new adapters be out of process. It will decouple your code from istio. The documentation of your adapter will still be available on istio.io.

It absolves istio from code reviewing adapters and also acceptIng vendor specific code. It will also future proof you from changes to Mixer.

We will certainly help you with this process.

Thanks.

**Comment by osswangxining** [comment]
@geeknoid @mandarjog   So far we move this adapter to Out of Process type and put the codes into new repository:  https://github.com/osswangxining/alicloud-istio-grpcadapter .  Any your comments? Thanks.

One question is how to submit the documentation of this adapter on istio.io as you mentioned that the doc will still be available on istio.io. @mandarjog

**Comment by geeknoid** [comment]
To publish your adapter to istio.io, you need to have a GitHub repo where
the auto-generated .html file can be found. Once you have that, you submit
a PR to change this file and add yourself in there:

https://github.com/istio/istio.github.io/blob/master/scripts/grab_reference_docs.sh#L36

Right now, this is how we pull in the Apigee adapter docs for example. Now
that this is likely to happen more often, I'll change that script such that
it is a one-line change to add a new adapter's repo and have it scanned for
content.

The script is not currently run via CI, we run this by hand once in a
while. So if you change the script, it might take a week or two before the
change is reflected on preliminary.istio.io. And it'll take until the next
release before it gets reflected on istio.io

Thanks.

**Comment by osswangxining** [comment]
Thanks for your review. And we moved the adapter to out of process which is kept in another repository:  https://github.com/osswangxining/alicloud-istio-grpcadapter.git

Resubmit another PR: https://github.com/istio/istio.github.io/pull/2301 
can you help review again? thanks.

**Comment by mandarjog** [comment]
You are using istio.io/istio/mixer/adapter/cloudmonitor in your imports. That does not exist in the repo. You need to define this in your adapter repo.

**Comment by osswangxining** [comment]
fix in commit https://github.com/osswangxining/alicloud-istio-grpcadapter/commit/3853d0e3f5a918471edd044625c1c3cf6162aa97

**Comment by mandarjog** [comment]
Thanks @osswangxining .
Once you are ready, we would love to see a demo in the telemetry and policy workgoup call.
Next one is on 29th of August at 11:00AM US Pacific Time.

**Comment by mandarjog** [comment]
I am closing the PR, since we have addressed this in other ways.

**Pull request closed without merging**